### PR TITLE
(PUP-9389) Add staging_location parameter to File

### PR DESCRIPTION
**Pull request by bkochendorfer**
## Description of Changes
When Puppet creates a new file it creates it in a temporary location. This location is currently the location the file will exist in plus a unique name, eg /etc/foo123fsd for /etc/foo. In cases where a user wants to first render a file to a temporary staging location, perform some validation commands and then move the file to the final location.

This avoids issues where, for example, an application reads all the files in a directory when performing a task (sudo) or an application reloads configs automatically from a directory.

## Open Questions
It is unclear to me if I need to do anything different in the case where Puppet is running on Windows and we want to render first to a temporary staging location. 

(edit) Oh nice, the tests run on Windows and looks to work.

## Jira
https://tickets.puppetlabs.com/browse/PUP-9389

### CODE REVIEW

**Comment by shrug** [comment]
There is some discussion in the ticket at https://tickets.puppetlabs.com/browse/PUP-9389

File: lib/puppet/util.rb [hunk 1]
```diff
@@ -551,19 +551,22 @@ def self.pretty_backtrace(backtrace = caller(1))
...
   # or a symbolic file mode.
+  #
+  # The staging_location is a location to render the temporary file before
+  # moving the file to it's final location.
 
   DEFAULT_POSIX_MODE = 0644
   DEFAULT_WINDOWS_MODE = nil
 
-  def replace_file(file, default_mode, &block)
+  def replace_file(file, default_mode, staging_location, &block)
```
Comment (joshcooper) [inline lib/puppet/util.rb]: To maintain backwards compatibility, this should be `staging_location = nil` so callers that expect the previous API don't break. Then we only need to update the one call site where we want to override `staging_location`.

**Comment by bkochendorfer** [comment]
This seems to be failing in [JRuby](https://travis-ci.org/puppetlabs/puppet/jobs/484032730#L1453-L1454) looks to be pretty close:
```
  1) Puppet::Util::Storage when storing to the state file does not expire entries when statettl is 'unlimited'
     Failure/Error:
       expect(Puppet::Util::Storage.state).to eq(
         {
           :yayness => {
             :checked => recent_checked
           },
           :older => {
             :checked => older_checked
           }
         }
       )
       expected: {:older=>{:checked=>2018-10-01 01:15:54.188781000 +0000}, :yayness=>{:checked=>2019-01-24 19:02:34.188776000 +0000}}
            got: {:older=>{:checked=>2018-10-01 01:15:54.188781000 +0000}, :yayness=>{:checked=>2019-01-24 19:02:34.188775999 +0000}}
       (compared using ==)
       Diff:
       @@ -1,3 +1,3 @@
        :older => {:checked=>2018-10-01 01:15:54.188781000 +0000},
       -:yayness => {:checked=>2019-01-24 19:02:34.188776000 +0000},
       +:yayness => {:checked=>2019-01-24 19:02:34.188775999 +0000},
     # ./spec/unit/util/storage_spec.rb:324:in `block in (root)'
     # util/rspec_runner:44:in `run'
     # util/rspec_runner:59:in `<main>'
Finished in 1 minute 18.96 seconds (files took 8.91 seconds to load)
```

**Comment by jhelwig** [comment]
@bkochendorfer Still poking at the jRuby failure, or did you want a hand dealing with it? It looks completely unrelated to your changes from just a fragile test that should be updated to not look at the full timestamp.

**Comment by bkochendorfer** [comment]
@jhelwig No I did not investigate as it did not seem related to my changes here. Posted the failure so folks could point me in the right direction if it was.

**Comment by kris-bosland** [comment]
@bkochendorfer , we fixed this flakey test on master, if you rebase it should pass.  Thanks!

**Comment by bkochendorfer** [comment]
One test seems to have had a segfault. Is there a way to kick the build again? Or is it related to something I changed?
```
bundle : C:/projects/puppet/lib/puppet/util/windows/principal.rb:115: [BUG] Segmentation fault
At line:14 char:1
+ bundle exec rake parallel:spec[2]
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (C:/projects/pup...mentation fault:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
ruby 2.5.3p105 (2018-10-18 revision 65156) [x64-mingw32]
```

**Comment by jhelwig** [comment]
The segfault on Windows is nothing related to any of your changes. It's just something that happens occasionally on AppVeyor. Not entirely sure why.

File: spec/unit/util_spec.rb [hunk 2]
```diff
@@ -871,6 +871,23 @@ def withenv_utf8(&block)
...
+      temp_target = Puppet::FileSystem::Uniquefile.new(Puppet::FileSystem.basename_string(new_target), Dir.tmpdir())
+
+      begin
+        expect(Puppet::FileSystem::Uniquefile).to receive(:new)
+          .with(Puppet::FileSystem.basename_string(new_target), Dir.tmpdir())
+          .and_return(temp_target)
+        subject.replace_file(new_target, 0555, Dir.tmpdir()) {|fh| fh.puts "foo" }
+      ensure
+        Puppet::FileSystem.unlink(new_target) if Puppet::FileSystem.exist?(new_target)
+        Puppet::FileSystem.unlink(temp_target) if Puppet::FileSystem.exist?(temp_target.to_s)
```
Comment (jhelwig) [inline spec/unit/util_spec.rb]: If I'm correctly understanding what this test is trying to do, I think we can simplify it to something like the following?

```ruby
    it "should use a temporary staging location if provided" do
      new_target = File.join(tmpdir('new_file'), 'new_file.baz')
      staging_target = tmpdir('staging_file')

      subject.replace_file(new_target, 0555, staging_target) do |fh|
        expect(File.dirname(fh.path)).to eq(staging_target)

        fh.puts "foo"
      end
    end
```

Comment (jhelwig) [reply spec/unit/util_spec.rb]: The `ensure` block that's removing the temp files shouldn't be necessary any more. Part of what `tmpdir` does is add the directories it creates to a list that gets cleaned up at the end of each test.

**Comment by jhelwig** [comment]
@bkochendorfer Looks good to me. Do you want a chance to squash those last two cleanup commits into the one that creates the test before I merge things?

**Comment by bkochendorfer** [comment]
Thank you @jhelwig, squished!

**Comment by jhelwig** [comment]
@bkochendorfer Thanks for all your patience, and work on this!

**Pull request merged**